dmn-engine/src/main/scala/org/camunda/dmn/DmnEngine.scala

```diff
@@ -194,17 +193,15 @@ class DmnEngine(configuration: DmnEngine.Configuration =
       context: EvalContext): Either[Failure, EvalResult] = {
     decisionEval
       .eval(decision, context)
-      .right
-      .map(result => {
-        val log = new AuditLog(context.dmn, context.auditLog.toList)
-
-        auditLogListeners.map(_.onEval(log))
-
-        result match {
-          case ValNull => NilResult
-          case result  => Result(valueMapper.unpackVal(result))
-        }
-      })
+      .map {
+        case ValNull => NilResult
+        case result  => Result(valueMapper.unpackVal(result))
+      } match {
+      case anyResult =>
+        val log = AuditLog(context.dmn, context.auditLog.toList)
+        auditLogListeners.foreach(_.onEval(log))
```

Comment:
The listener is invoked if the evaluation is successful or not. But it is hard to distinguish between the cases. 

We need a second method in the listener for the case that the evaluation failed. Some implementations, like the integration for Camunda, will behave differently (i.e. don't persist the history).

Later, I want to change the API a bit. Instead of returning only on the decision result or the failure, it should return the audit log as well. That would make it easier to keep the result and the audit log together. And it would enable some use cases (i.e. decision tree).

Reply:
Something like this:
```
  trait AuditLogListener {

    def onEval(log: AuditLog)
    def onFailedEval(log: AuditLog)

  }
```

>  it should return the audit log as well. 

I aggree. That is exactly what I had to implement in my project - in my case for reporting.

Reply:
Yes :+1: 

Maybe, we name the new method simply `onFailure(log: AuditLog)`.

Reply:
@saig0 thanks for your answers and suggestions!

I waited to simplify all the Evaluators - as I was not sure if you liked the audit function.

I fixed all your suggestions - thanks for your time - looks better now.

Concerning the `trait AuditLogListener`. To add the additional function would break compatibility - do you really want this for now?

> Note that I want to bring the PR in a good final state before merging it. Currently, Camunda evaluates if we can use the DMN engine in the platform 🎈

That would be great! More Scala code in Camunda:laughing:.